Feature/shopping cart

lesser1 opened the pull request:
For what has been done, please see the commit messages.

Remark: When a user attempts to visit the shopping cart route but is not allowed to access the page, I would like to discuss, design, and implement that as a separate issue. This is because, in general, redirects still need to be implemented for the frontend.

[Gropius Issue...](https://frontend.gropius.duckdns.org/components/9d12d6ac-e85a-407c-bdff-a8321fb75d3a/issues/fe3cb70c-69fe-4939-bdc6-895c27caf9b0)

### Definition of Done
- [x] Requirements of the issue are met
- [x] Code has been reviewed
- [x] Code is documented
- [x] GraphQL related artefacts are documented

lesser1 committed:
Modify error handler:

    - Add synchronous version of
      function pushErrorNotificationIfNecessary (old name).
    - Rename pushErrorNotificationIfNecessary to
      awaitActionAndPushErrorIfNecessary to indicate async mode.

lesser1 committed:
Add shopping cart to the app store:

    - Add two interfaces representing a shopping cart:
      shoppingCart and shoppingCartItem
    - Add shopping cart as state to the app store.
    - Add function that resets stateful information related to the
      current user.

lesser1 committed:
Restore shopping cart after login:

    - Load shopping cart of current user.
    - Add getter to app store that determines whether or not the
      shopping cart is enabled.
    - Hide shopping cart functionality if user is not buyer,
      disable it if user is buyer but not logged in.

lesser1 committed:
Feature: User can add product v. to the cart

    - Add two GraphQL mutations to shoppingcart.graphql for adding and
      updating shopping cart items.
    - Make SHOPPING CART toolbar button log shopping cart if dev env.
      This for testing purposes only.
    - Add strings to errorMessages.ts
    - Implement ADD TO CART button in ProductView

lesser1 committed:
Provide doc comment to GraphQL mutation

lesser1 committed:
Modify shoppingcart.graphql:

    - Modify query getShoppingCartOfUser to also return the ID of the
      product, the name of the current version and the retail price.
    - Add delete mutation.

lesser1 committed:
Extend ShoppingCartItem interface

    - Add properties to ShoppingCartItem interface.
    - Add two async functions to the app store for managing the user's
      shopping cart: update and delete
    - Add error messages

lesser1 committed:
Modify ProductSummary:

From now on, a product summary can optionally represent a shopping
cart item.

lesser1 committed:
Add view / page for the shopping cart:

    - Add ShoppingCartView.vue.
    - Add seperate SFC: ShoppingCartToolbar.vue Because the toolbar is
      used twice in the ShoppingCartView.vue.
    - Add ShoppingCartView to the router.

lesser1 committed:
Run linter and formater...

lesser1 commented:
PR #38 should be merged first.

lesser1 committed:
Merge branch 'main' into feature/shopping-cart

nk-coding commented on src/components/ProductSummary.vue:
```diff
@@ -1,32 +1,76 @@
+                    <ProductPrice :price="price" :retail-price="retailPrice" />
+                </div>
+            </div>
+            <div v-if="isShoppingCartItem" class="d-flex ga-2">
```
tbh this feels a bit like this component serves two purposes: just displaying it and also this shopping cart thing
generally speaking, this violates separation of concerns a bit
I personally would probably have added a slot to this component, which by default is empty, but where for the shopping cart use case, the new stuff can be added, but feel free to ignore this if you deem it too much work

lesser1 replied:
How about I open a Gropius Issue for this so that I can change things later? Because I think there a more pressing issues right now.

nk-coding replied:
fine for me

nk-coding commented on src/components/ProductSummary.vue:
```diff
@@ -38,28 +82,93 @@ const props = defineProps({
+ */
+const emit = defineEmits<{
+    (
+        event: 'count-changed',
```
why not use a model for this? you can define a custom one using https://vuejs.org/api/sfc-script-setup.html#definemodel
then in the parent, you can just use v-model and do not have to handle the event differently (disclaimer: I did not yet look how you use this event)

lesser1 replied:
Same as above: How about I open a Gropius Issue for this so that I can change things later? Because I think there a more pressing issues right now. I have to say that I like both suggestions -- slot and custom model -- but I already put a lot of time into this issue here...

nk-coding replied:
this should be like 5 LOC to change
but if you want, fine for me

nk-coding commented on src/model/shoppingCart.ts:
```diff
@@ -0,0 +1,10 @@
+ * @interface
+ */
+export interface ShoppingCart {
+    lastUpdatedAt: string | null
```
field documentations missing

lesser1 replied:
fa164861fb05f8c7fa5b900dab45ff2434fd4d54

nk-coding commented on src/model/shoppingCartItem.ts:
```diff
@@ -0,0 +1,14 @@
+ * @interface
+ */
+export interface ShoppingCartItem {
+    id: string
```
field documentations missing

lesser1 replied:
fa164861fb05f8c7fa5b900dab45ff2434fd4d54

nk-coding commented on src/store/app.ts:
```diff
@@ -206,6 +231,53 @@ export const useAppStore = defineStore('app', {
+                    shoppingCartOfCurrentUser.lastUpdatedAt
+                this.shoppingCart.items = []
+
+                shoppingCartOfCurrentUser.shoppingcartItems.nodes.forEach(
```
soo, Becker has this 20 LOC rule for function bodies
while I guess this rule is not that strict for frontend dev, this is still a bit long
maybe this foreach block in the middle can be refactored into another function?

lesser1 replied:
5d5a562aea9175c9934346122284a7d9950a6048

nk-coding commented on src/store/app.ts:
```diff
@@ -301,5 +382,136 @@ export const useAppStore = defineStore('app', {
+         * @param productVariantId - The ID of the product variant to add.
+         * @param count - The quantity of the product variant to add.
+         */
+        async addProductVariantToShoppingCart(
```
same as above, maybe refactoring into multiple functions is reasonably possible

lesser1 replied:
89fda6b208dbabdbc3f6e13b391b53f5b1e889ab

nk-coding commented on src/store/app.ts:
```diff
@@ -301,5 +382,136 @@ export const useAppStore = defineStore('app', {
+
+            let nameOfProductVariant: string
+
+            const existingShoppingCartItems = this.shoppingCart.items.filter(
```
as you only need the (optional) first later, you could also use `find`

lesser1 replied:
89fda6b208dbabdbc3f6e13b391b53f5b1e889ab

lesser1 committed:
Add missing JSDoc documentation of properties

lesser1 committed:
Refactor functions RE restoring the shopping cart

lesser1 committed:
Refactor functions RE shopping cart management:

    - Add separate file for stateless functions:
      shoppingCartManagement.ts
    - Extract smaller functions.
    - Rename functions.

nk-coding reviewed (changes requested):
Bug: if I add sth to the shopping card in buyer mode, and then refresh the frontend (I am in admin mode then), and then change the mode to buyer again and access the shopping cart, the items are never fetched and therefore not there

lesser1 committed:
Fix bug RE restoring of shopping cart:

    - Move emptyShoppingCart const to shoppingCartManagement.ts.
    - Introduce helper action emptyTheShoppingCart.
    - Extend logic of action restoreTheShoppingCart:
        - Empty shopping cart at first -- as a fallback
        - Then query the shopping cart.
    - Add watcher in App.vue to restore shopping cart if necessary.

lesser1 commented:
> Bug: if I add sth to the shopping card in buyer mode, and then refresh the frontend (I am in admin mode then), and then change the mode to buyer again and access the shopping cart, the items are never fetched and therefore not there

8b91be771752460437cc95c16e9f624271d2960d

lesser1 committed:
Merge branch 'main' into feature/shopping-cart

    - Apply renaming of errorHandler.ts function in RestockDialog.vue
    - Regenerate generated.ts via 'yarn codegen'.

lesser1 committed:
Fix: Items would disappear after changes to them

    - Add key prop to list item (best practice).
    - Modify restoreTheShoppingCart function in app.ts:
        Only empty the shopping cart if it cannot be queried and
        updated correctly.

nk-coding reviewed (approved):
bug seems to be fixed
